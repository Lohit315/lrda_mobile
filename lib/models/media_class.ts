import { MediaData } from "../../types";

export class Media {
  uuid: string;
  type: string;
  uri: string;

  constructor({ uuid, type, uri }: MediaData) {
    this.uuid = uuid;
    this.type = type;
    this.uri = uri;
  }

  getUuid(): string {
    return this.uuid;
  }

  getType(): string {
    return this.type;
  }

  getUri(): string {
    return this.uri;
  }
}

export class AudioType extends Media {
  duration: string;
  name: string;

  constructor({
    uuid,
    uri,
    duration,
    name,
  }: MediaData & { duration: string; name: string }) {
    super({ uuid, type: "audio", uri });
    this.duration = duration;
    this.name = name;
  }

  isAudio(): boolean {
    return true;
  }

  getDuration(): string {
    return this.duration;
  }
}

export class VideoType extends Media {
  thumbnail: string;
  duration: string;

  constructor({
    uuid,
    uri,
    thumbnail,
    duration,
  }: MediaData & { thumbnail: string; duration: string }) {
    super({ uuid, type: "video", uri });
    this.thumbnail = thumbnail;
    this.duration = duration;
  }

  getDuration(): string {
    return this.duration;
  }

  getThumbnail(): string {
    return this.thumbnail;
  }
}

export class PhotoType extends Media {
  constructor({ uuid, uri }: MediaData) {
    super({ uuid, type: "image", uri });
  }
}
