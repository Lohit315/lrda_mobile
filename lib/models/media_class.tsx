type MediaData = {
    uuid: string;
    type: string;
    uri: string;
    thumbnail: string;
  };
  
  export class Media {
    uuid: string;
    type: string;
    uri: string;
    thumbnail: string;
  
    constructor({ uuid, type, uri, thumbnail }: MediaData) {
      this.uuid = uuid;
      this.type = type;
      this.uri = uri;
      this.thumbnail = thumbnail;
    }
  
    getUuid(): string {
      return this.uuid;
    }
  
    getType(): string {
      return this.type;
    }
  
    getUri(): string {
      return this.uri;
    }
  
    getThumbnail(): string {
      return this.thumbnail;
    }
  
    isVideo(): boolean {
      return this.type === 'video';
    }
  
    isImage(): boolean {
      return this.type === 'image';
    }
  }
  