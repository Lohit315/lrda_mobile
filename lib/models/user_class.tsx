import { UserData } from "../../types";

export class User {
  private static instance: User;
  private userData: UserData | null = null;

  private constructor() {}

  public static getInstance(): User {
    if (!User.instance) {
      User.instance = new User();
    }
    return User.instance;
  }

  public async login(username: string, password: string): Promise<string> {
    try {
      const response = await fetch(
        "http://lived-religion-dev.rerum.io/deer-lr/login",
        {
          method: "POST",
          mode: "cors",
          cache: "no-cache",
          headers: {
            "Content-Type": "application/json;charset=utf-8",
          },
          body: JSON.stringify({
            username: username,
            password: password,
          }),
        }
      );

      if (response.ok) {
        this.userData = await response.json();
        return "success";
      } else {
        throw new Error("There was a server error logging in.");
      }
    } catch (error) {
      console.error(error);
      return Promise.reject(error);
    }
  }

  public async logout() {
    try {
      await fetch("http://lived-religion-dev.rerum.io/deer-lr/logout", {
        method: "POST",
        mode: "cors",
        headers: {
          "Content-Type": "text/plain",
        },
      })
        .then((response) => {
          if (response.ok) {
            this.userData = null;
            console.log('User logged out');
          }
        })
        .catch((err) => {
          return err;
        });
    } catch (error) {console.log('User did not succesfully log out')}
  }

  public getId(): string | null {
    return this.userData?.["@id"] ?? null;
  }

  public getName(): string | null {
    return this.userData?.name ?? null;
  }

  public getRoles(): { administrator: boolean; contributor: boolean } | null {
    return this.userData?.roles ?? null;
  }
}
