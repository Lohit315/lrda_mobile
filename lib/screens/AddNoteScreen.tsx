import React, { useEffect, useState, useRef } from "react";
import {
  Alert,
  View,
  StyleSheet,
  TextInput,
  TouchableOpacity,
  Platform,
} from "react-native";
import { Note, AddNoteScreenProps } from "../../types";
import PhotoScroller from "../components/photoScroller";
import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import { User } from "../models/user_class";
import { Ionicons } from "@expo/vector-icons";
import { Media, AudioType } from "../models/media_class";
import AudioContainer from "../components/audio";
import * as Location from "expo-location";
import ApiService from "../utils/api_calls";
import TagWindow from "../components/tagging"

const user = User.getInstance();

const AddNoteScreen: React.FC<AddNoteScreenProps> = ({ navigation, route }) => {
  const [titleText, setTitleText] = useState("");
  const [bodyText, setBodyText] = useState("");
  const [newMedia, setNewMedia] = useState<Media[]>([]);
  const [newAudio, setNewAudio] = useState<AudioType[]>([]);
  const [tags, setTags] = useState<string[]>([]);
  const [viewMedia, setViewMedia] = useState(false);
  const [viewAudio, setViewAudio] = useState(false);
  const [isTagging, setIsTagging] = useState(false);
  const [isPublished, setIsPublished] = useState(false);
  const [location, setLocation] = useState<{
    latitude: number;
    longitude: number;
  } | null>(null);

  useEffect(() => {
    console.log("new Media array:", newMedia);
  }, [newMedia]);

  useEffect(() => {
    (async () => {
      let { status } = await Location.requestForegroundPermissionsAsync();
      if (status !== "granted") {
        console.log("Permission to access location was denied");
        return;
      }

      let location = await Location.getCurrentPositionAsync({});
      setLocation({
        latitude: location.coords.latitude,
        longitude: location.coords.longitude,
      });
    })();
  }, []);

  const saveNote = async () => {
    try {
      const newNote = {
        title: titleText,
        text: bodyText,
        media: newMedia,
        audio: newAudio,
        creator: user.getId(),
        latitude: location?.latitude.toString() || "",
        longitude: location?.longitude.toString() || "",
        published: isPublished,
        tags: tags,
      };
      const response = await ApiService.writeNewNote(newNote);

      const obj = await response.json();
      const id = obj["@id"];

      route.params.refreshPage();
      navigation.goBack();
    } catch (error) {
      console.error("An error occurred while creating the note:", error);
    }
  };

  return (
    <View>
      <View style={styles.topContainer}>
        <TouchableOpacity style={styles.topButtons} onPress={saveNote}>
          <Ionicons name="arrow-back-outline" size={30} color="white" />
        </TouchableOpacity>
        <TextInput
          style={styles.title}
          placeholder="Title Field Note"
          onChangeText={(text) => setTitleText(text)}
          value={titleText}
        />

        {isPublished ? (
          <TouchableOpacity
            style={styles.topButtons}
            onPress={() => setIsPublished(!isPublished)}
          >
            <Ionicons name="earth" size={30} color="white" />
          </TouchableOpacity>
        ) : (
          <TouchableOpacity
            style={styles.topButtons}
            onPress={() => setIsPublished(!isPublished)}
          >
            <Ionicons name="earth-outline" size={30} color="white" />
          </TouchableOpacity>
        )}
      </View>
      <View style={styles.keyContainer}>
          <TouchableOpacity onPress={() =>{
            setViewMedia(!viewMedia);
            setViewAudio(false);
            setIsTagging(false);
          }}>
            <Ionicons name="images-outline" size={30} color="black" />
          </TouchableOpacity>
          <TouchableOpacity onPress={() =>{
            setViewMedia(false);
            setViewAudio(!viewAudio);
            setIsTagging(false);
          }}>
            <Ionicons name="mic-outline" size={30} color="black" />
          </TouchableOpacity>
          <TouchableOpacity>
            <Ionicons name="location-outline" size={30} color="black" />
          </TouchableOpacity>
          <TouchableOpacity>
            <Ionicons name="time-outline" size={30} color="black" />
          </TouchableOpacity>
          <TouchableOpacity onPress={() =>{
            setViewMedia(false);
            setViewAudio(false);
            setIsTagging(!isTagging);
          }}>
            <Ionicons name="pricetag-outline" size={30} color="black" />
          </TouchableOpacity>
        </View>
      <View style={styles.container}>
        <KeyboardAwareScrollView
          nestedScrollEnabled
          showsVerticalScrollIndicator={false}
          style={{ overflow: "hidden", paddingTop: 10, paddingBottom: 100 }}
        >
          {viewMedia && (
            <PhotoScroller newMedia={newMedia} setNewMedia={setNewMedia} />
          )}
          {viewAudio && (
            <AudioContainer newAudio={newAudio} setNewAudio={setNewAudio} />
          )}
          {isTagging && (
          <TagWindow tags={tags} setTags={setTags} />
          )}
          <View style={styles.inputContainer}>
            <TextInput
              style={styles.input}
              placeholder="Write your note here"
              multiline={true}
              textAlignVertical="top"
              onChangeText={(text) => setBodyText(text)}
              value={bodyText}
            />
          </View>
        </KeyboardAwareScrollView>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  topContainer: {
    flex: 1,
    justifyContent: "space-between",
    paddingHorizontal: 5,
    minHeight: "15%",
    paddingTop: "15%",
    flexDirection: "row",
    backgroundColor: "#F4DFCD",
    alignItems: "center",
    textAlign: "center",
  },
  topText: {
    flex: 1,
    maxWidth: "100%",
    fontWeight: "700",
    fontSize: 32,
    textAlign: "center",
  },
  topButtons: {
    backgroundColor: "#111111",
    borderRadius: 50,
    width: 50,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
    zIndex: 99,
  },
  toggles: {
    backgroundColor: "#111111",
    borderRadius: 50,
    width: 50,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
    marginLeft: 10,
    zIndex: 99,
  },
  container: {
    paddingHorizontal: 16,
    backgroundColor: "white",
    overflow: "hidden",
    paddingBottom: "50%",
  },
  title: {
    height: 45,
    width: "70%",
    borderColor: "#111111",
    borderWidth: 1,
    borderRadius: 30,
    paddingHorizontal: 10,
    textAlign: "center",
    fontSize: 30,
  },
  input: {
    flex: 1,
    borderColor: "#111111",
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    fontSize: 22,
    paddingBottom: "90%",
  },
  addButton: {
    position: "absolute",
    bottom: 20,
    right: 20,
    backgroundColor: "#111111",
    borderRadius: 50,
    width: 50,
    height: 50,
    alignItems: "center",
    justifyContent: "center",
  },
  keyContainer: {
    height: 60,
    paddingVertical: 5,
    width: "100%",
    backgroundColor: "#F4DFCD",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    paddingHorizontal: 40,
  },
  saveText: {
    color: "#111111",
    fontWeight: "bold",
    fontSize: 12,
  },
  inputContainer: {
    height: 400,
    justifyContent: "space-between",
  },
});

export default AddNoteScreen;
