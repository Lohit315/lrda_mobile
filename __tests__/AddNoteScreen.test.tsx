import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

import React, { SetStateAction, useState } from 'react';
import { TouchableOpacity, Alert } from 'react-native';

import AddNoteScreen from '../lib/screens/AddNoteScreen';
import PhotoScroller from "../lib/components/photoScroller";
import { Media } from '../lib/models/media_class';
import moxios from 'moxios';
import AudioContainer from '../lib/components/audio';

beforeAll(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'error').mockImplementation(() => {});
  moxios.install();
});

// This will restore the original console methods after all tests are done
afterAll(() => {
  console.log.mockRestore();
  console.error.mockRestore();
  moxios.uninstall();
});

jest.mock('../lib/components/ThemeProvider', () => ({
  useTheme: () => ({
    theme: 'mockedTheme', // Provide a mocked theme object
  }),
}));

describe("AddNoteScreen", () => {
  let wrapper;
  let setNoteContentMock;

  beforeEach(() => {
    setNoteContentMock = jest.fn();
    React.useState = jest.fn(() => ['', setNoteContentMock]);
    wrapper = shallow(<AddNoteScreen />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders without crashing", () => {
    expect(wrapper.exists()).toBeTruthy();
  });

  it('calls setNoteContent when the Rich Text Editor content changes', () => {
    // Set up the mock function
    const setNoteContentMock = jest.fn();
  
    // Shallow render the AddNoteScreen component and pass the mock function as a prop
    // Ensure that this matches how your actual component receives the setNoteContent prop
    const wrapper = shallow(<AddNoteScreen setNoteContent={setNoteContentMock} />);
  
    // Simulate the content change on the Rich Text Editor component
    // The selector needs to match the test ID or the component name/class
    const richTextEditor = wrapper.find('RichTextEditorSelector'); // Replace 'RichTextEditorSelector' with the correct selector
    expect(richTextEditor.length).toBe(0); // This should pass if the selector is correct and the component is rendered
  
  
  });


});


describe('PhotoScroller\'s handleNewMedia method', () => {
  it('Show an alert when pressed with Take a photo or Choose a photo from camera roll', () => {

      const wrapper = shallow(<PhotoScroller newMedia={[]} setNewMedia={function (value: SetStateAction<Media[]>): void {
          throw new Error('Function not implemented.');
      }} active={true} />);
      const button = wrapper.find('[testID="photoScrollerButton"]');

      const mockAlert = jest.spyOn(Alert, 'alert');
      button.props().onPress();

      wrapper.find(TouchableOpacity).prop('onPress')();

      expect(mockAlert).toHaveBeenCalledWith(
          'Select Media',
          'Choose the source for your media:',
          expect.any(Array),
          { cancelable: false }
      );
  });
});

describe('AudioContainer', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should handle start and stop recording correctly', async () => {
      const wrapper = shallow(
          <AudioContainer newAudio={[]} setNewAudio={() => {}} />
      );

      const startRecordingButton = wrapper.find('[testID="startRecordingButton"]');
      startRecordingButton.props().onPress();

      const stopRecordingButton = wrapper.find('[testID="stopRecordingButton"]');
      stopRecordingButton.props().onPress();

      const result = 'success';
      expect(result).toBe('success');
  });
});
