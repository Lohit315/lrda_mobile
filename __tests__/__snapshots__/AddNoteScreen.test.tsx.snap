// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNoteScreen renders without crashing 1`] = `ShallowWrapper {}`;
